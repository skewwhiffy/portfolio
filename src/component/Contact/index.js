import React, { useState } from 'react'

const Contact = () => {
  const [form, setForm] = useState({
    name: '',
    email: '',
    message: ''
  })
  const [errorMessage, setErrorMessage] = useState('')
  const [valid, setValid] = useState(false)

  const isEmail = email => {
    const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
    return re.test(String(email).toLowerCase())
  }

  const handleChange = event => {
    const { name, value } = event.target
    const newForm = {
      ...form,
      [name]: value
    }
    setForm(newForm)
    if (!value) {
      setErrorMessage(`Field '${name}' is required`)
      return
    }
    if (name === 'email' && !isEmail(value)) {
      setErrorMessage('Email address is invalid')
      return
    }
    setErrorMessage('')
  }

  const handleSubmit = event => {
    event.preventDefault()
    console.log(form)
  }

  const validate = event => {
    const { name, value } = event.target
    const newForm = {
      ...form,
      [name]: value
    }
    setValid(!!newForm.name && !!newForm.message && isEmail(newForm.email))
  }

  const { name, email, message } = form
  return (
    <div className="contact">
      <h1>Contact</h1>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="name">Name:</label>
          <input placeholder="name" type="text" name="name" onChange={validate} onBlur={handleChange} defaultValue={name} />
        </div>
        <div>
          <label htmlFor="email">Email address:</label>
          <input placeholder="email" type="email" name="email" onChange={validate} onBlur={handleChange} defaultValue={email} />
        </div>
        <div>
          <label htmlFor="message"> Message:</label>
          <textarea placeholder="message" name="message" rows="5" onChange={validate} onBlur={handleChange} defaultValue={message} />
        </div>
        {errorMessage && (
          <div>
            <p className="error-text" data-testid="errorText">{errorMessage}</p>
          </div>
        )}
        <button disabled={!valid} type="submit" data-testid="submitButton">Submit</button>
      </form>
    </div>
  )
}

export default Contact