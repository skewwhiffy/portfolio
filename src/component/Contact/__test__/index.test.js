import React from 'react';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import Contact from '..';

afterEach(cleanup)

describe('Contact component', () => {
  let asFragment
  let getByPlaceholderText
  let getByTestId
  let nameElement
  let emailElement
  let messageElement
  let submitButton

  const getErrorTextElement = () => getByTestId('errorText')

  beforeEach(() => {
    ({ asFragment, getByPlaceholderText, getByTestId } = render(<Contact />))
    nameElement = getByPlaceholderText('name')
    emailElement = getByPlaceholderText('email')
    messageElement = getByPlaceholderText('message')
    submitButton = getByTestId('submitButton')
  })

  it('matches snapshot DOM node structure', () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it('informs of empty name', () => {
    nameElement.focus()
    nameElement.blur()

    expect(getErrorTextElement()).toHaveTextContent('name')
  })

  it('informs of empty email', () => {
    emailElement.focus()
    emailElement.blur()

    expect(getErrorTextElement()).toHaveTextContent('email')
  })

  it('informs of invalid email', () => {
    emailElement.focus()
    userEvent.type(emailElement, 'not_valid')
    emailElement.blur()

    expect(getErrorTextElement()).toHaveTextContent('invalid')
  })

  it('informs of empty message', () => {
    messageElement.focus()
    messageElement.blur()

    expect(getErrorTextElement()).toHaveTextContent('message')
  })

  it('enables submit button as soon as form is valid', () => {
    expect(submitButton.disabled).toBeTruthy()

    userEvent.type(nameElement, 'John Coltrane')
    expect(submitButton.disabled).toBeTruthy()

    userEvent.type(emailElement, 'john.coltrane@test.com')
    expect(submitButton.disabled).toBeTruthy()

    userEvent.type(messageElement, 'Play more bebop, it\'s really good for the soul.')
    expect(submitButton.disabled).toBeFalsy()
  })
})