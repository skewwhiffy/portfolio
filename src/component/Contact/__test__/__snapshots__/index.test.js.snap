// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component matches snapshot DOM node structure 1`] = `
<DocumentFragment>
  <div
    class="contact"
  >
    <h1>
      Contact
    </h1>
    <form>
      <div>
        <label
          for="name"
        >
          Name:
        </label>
        <input
          name="name"
          placeholder="name"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email address:
        </label>
        <input
          name="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="message"
        >
           Message:
        </label>
        <textarea
          name="message"
          placeholder="message"
          rows="5"
        />
      </div>
      <button
        data-testid="submitButton"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
